src/hotspot/share/code/dependencyContext.cpp

```diff
@@ -205,30 +205,47 @@ void DependencyContext::clean_unloading_dependents() {
   }
 }
 
+nmethodBucket* DependencyContext::release_and_get_next_not_unloading(nmethodBucket* b) {
+     nmethodBucket* next = b->next_not_unloading();
+     release(b);
+     return next;
+ }
+
 //
 // Invalidate all dependencies in the context
-int DependencyContext::remove_all_dependents() {
+void DependencyContext::remove_all_dependents() {
   nmethodBucket* b = dependencies_not_unloading();
-  set_dependencies(NULL);
-  int marked = 0;
-  int removed = 0;
-  while (b != NULL) {
-    nmethod* nm = b->get_nmethod();
-    if (b->count() > 0 && nm->is_alive() && !nm->is_marked_for_deoptimization()) {
-      nm->mark_for_deoptimization();
-      marked++;
-    }
-    nmethodBucket* next = b->next_not_unloading();
-    removed++;
-    release(b);
-    b = next;
-  }
-  if (UsePerfData && removed > 0) {
-    _perf_total_buckets_deallocated_count->inc(removed);
-  }
-  return marked;
+   set_dependencies(NULL);
+   int removed = 0;
+   while (b != NULL) {
+     b = release_and_get_next_not_unloading(b);
```

Comment:
If we are being called by class unloading, finding an nmethod with nm->is_alive() returning true would be fatal, wouldn't it?  Could we make it clear this function is used for unloading, and add an assert that !nm->is_alive() && nm->has_flushed_dependencies()?

Reply:
> You are correct that nm should all be at least `unloading`, as if a klass is unloading all dependent `nmethods` should be unloading. So this code should boil down to:
> 
> ```c++
> void DependencyContext::remove_all_dependents() {
>   nmethodBucket* b = dependencies_not_unloading();
>    set_dependencies(NULL);
>    assert(b == nullptr, "All dependents should be unloading");
> }
> ```
> 
> As `dependencies_not_unloading()` will unlink and release every dependent that is `unloading`. Have run stress tests on that change and no violations were found. Will push and test this change.

From testing this seems not to be the case. Will investigate some more. So there are dependents that do no become `unloading` when the klass unloads.

Reply:
I would like to find out more in detail what you want to ensure, or situation you are afraid that might occur. 

I've tried to investigate what types of dependent nmethods are not `unloading` when the the klass with a dependency is unloaded. I suspect it has to do with class loading and compilation order. From what I understand the nmethod will only be unloading if it has a reference to a dead oop inside it. So the compiler must generate code which is dependent without having an oop baked in. 

I do find it difficult to reason about the lifetime of these dependencies (at least the logical part, the release, purge cycle of the memory seems pretty straight forward) and nmethods (the sweeper removal does make the nmethod lifecycle a lot easier to reason about tough). 

I am really interested in what you think. As this change is just always running an edge case which can happen in mainline. This just makes it default. Hotspot seems to have relied on this interaction being this way for some time. 

The windows machines finally woke up and the patch passed tier 1-7

Reply:
My concern was that we could be removing the dependencies from nmethods that are still using them.  But if that was the case, the old code would have the same problem.

Reply:
The compiler dependencies are protecting against loading of a class that is a subtype of X. If X is unloaded, then we can be pretty sure nobody is going to load a subclass of X.

Reply:
Isn't there something we can assert about the state of these nmethods?  Maybe unlinked and !in_use?

Reply:
Seems like the culprit it `CodeCache::UnloadingScope`. Currently the CodeCache::UnloadingScope does not encompass the klass unloading (different for each GC, some do some don't). Only the CodeCache unloading. This means that the is_unloading data is not up to date. 

After modifying serial to have the `UnloadingScope` covering both klass unloading and nmethod unloading this code seems to work.
```C++
void DependencyContext::remove_all_dependents() {
  nmethodBucket* b = dependencies_not_unloading();
  set_dependencies(NULL);
  assert(b == nullptr, "All dependents should be unloading");
}
```

The options here are to either fix the scopes for all GCs so it covers both, or introduce a slow `is_unloading` for the assert which does not use the cached unloading value. The second option still need to be GC aware as it requires the GCs notion of `is_alive`. The first option seems the best as it will be more correct with respect to the internal vm state. It will probably be better at catching future bugs as well.